Text support

prosif opened the pull request:
Add text support + support for floating point size/position values. Increases entity frame size from 8 indices to 46- original 8 plus 4 (+1 for two values after decimal for each size/position value), then 34 for text. 2 for x, y, then 32 for ASCII values of text. This can/will definitely be optimized, as not every frame needs to be so large since not every entity has text. But basic text support is here.

YazeedLoonat commented on src/Squisher.js:
```diff
@@ -52,16 +52,29 @@ class Squisher {
+            this.pixelBoard[entityFrameSize * i + 8] = Math.floor(this.entities[i].size.x * 100);
+            this.pixelBoard[entityFrameSize * i + 9] = this.entities[i].size.x * 10000 % 100;
+            this.pixelBoard[entityFrameSize * i + 10] = Math.floor(this.entities[i].size.y * 100);
+            this.pixelBoard[entityFrameSize * i + 11] = this.entities[i].size.y * 10000 % 100;
```
I'm not 1 hundo on this, 
looking at: 
https://www.khanacademy.org/computing/computer-science/cryptography/modarithmetic/a/modular-multiplication 
we get:
A * 10000 % 100 =  (A % 100 * 10000 % 100) % 100 =  (A % 100 * 0) % 100 = 0
Where A is any number

prosif commented:
I put up this PR as a rough v1 for myself, I didn't intend for another person to actually read it. But still, thanks for taking a look! 

I still need to add some stuff to this branch, including:

- text size
- text font
- optimization for entity frame size when they have no text

prosif commented:
I've updated this PR with a whole lot more stuff that includes support for SSL, audio, text & images.

This is still very much a preliminary PR, I just want to have all of the changes in one big thing with comments/changes over time. I will work on cleaning this PR up and fixing bugs, then moving on to the new data structure to store entity state in the squisher. 

Until then, this is pretty rad.

prosif commented:
And now I've read my (very rude sounding) previous comment, which reminds me that I need to add config for text as well. Sorry about that comment btw

prosif commented:
Alright so now I've made a couple more big changes. 

1. Removed SSL, as I realized I didn't need it anymore. Or at least I don't think so. Anyway it's a relief because I don't know how we'd set up SSL for every homegames instance. 
2. Replaced usage of `192..` with local hostname acquired from URL box
3. Added support for `homegames.link` to allow easy mapping to the local IP without having to physically type out your IP address (gross)
4. A bit of general cleanup with server stuff.

I still need to clean up what's here before it's ready to review & merge.

prosif commented:
Also, check out [homegames.link](https://github.com/prosif/homegames.link) to see what that is. You should be able to use it locally when testing @YazeedLoonat, let me know if it doesn't work. 

I've also updated http://picodeg.io to use this branch

prosif commented on src/Player.js:
```diff
@@ -1,29 +1,35 @@
     }
 
     disconnect() {
-        // handle cleanup here. will probably need to notify listeners. too many listeners.
```
Too true, past me. Too true.

prosif commented on src/Squisher.js:
```diff
@@ -93,13 +122,80 @@ class Squisher {
     }
+    
+    squish(entity) {
+        // Type (1) + Size (1) + color (4) + pos (4) + size (4) + text position (2) + text (32) + assets (37 * assetCount)
```
This function is gross and I don't like it. I want to clean this up but I'm still making big changes to what's sent over the socket so I will wait until that's a little more clear before abstracting this logic

prosif commented on src/games/draw.js:
```diff
@@ -1,17 +1,52 @@
     constructor() {
+        this.assets = {
+            "test": new Asset("url", {
+                "location": "https://www.nicepng.com/png/full/323-3239506_kanye-west-shrug-transparent.png"
```
nice

prosif commented on src/Squisher.js:
```diff
@@ -17,7 +19,39 @@ class Squisher {
+            
+            const assetType = gameAssets[key].info.type === "image" ? 1 : 2;
+
+            this.assets[key] = [ASSET_TYPE, assetType, encodedLength.charCodeAt(0), encodedLength.charCodeAt(1), encodedLength.charCodeAt(2), encodedLength.charCodeAt(3), ...assetKeyArray, ...payload];
```
This logic should be abstracted. Pretty sure we won't support video in the near future (though I suppose that's not too crazy), so image/audio are probably enough for assets for now

prosif commented on src/games/draw.js:
```diff
@@ -1,17 +1,49 @@
     constructor() {
+        this.assets = {
+            "test": new Asset("url", {
+                "location": "https://www.nicepng.com/png/full/323-3239506_kanye-west-shrug-transparent.png",
```
nice

prosif commented on web/App.js:
```diff
@@ -0,0 +1,235 @@
+        const frameType = buf[i];
+
+        if (frameType === 1) {
+            const assetType = buf[i + 1];
```
This is a weird combination of the inverse of the "squish" function in `Squisher.js` & asset decoding/storing. 

I would like to have one common module that can be shared by both the web app & the server that will squish/unsquish classes of thing. Kind of like a common "game node serializer". That's the intent of Squisher, really.

prosif commented on web/App.js:
```diff
@@ -0,0 +1,235 @@
+    unlock();
+});
+
+canvas.addEventListener('mouseup', function(e) {
```
Once this list of event listeners grows and starts to split up more based on device, we should abstract some of this stuff

prosif commented on web/App.js:
```diff
@@ -0,0 +1,235 @@
+    return event.key.length == 1 && event.key >= ' ' && event.key <= 'z' || event.keyCode >= 36 && event.keyCode <= 40;
+}
+
+document.addEventListener('keydown', function(e) {
```
TODO: Handle CMD+R or F5 to refresh.
